resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -417,8 +417,11 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
     }
   }
 
-  private def sparkContextInitialized(sc: SparkContext) = {
+  private def sparkContextInitialized(sc: SparkContext) = synchronized {
```

Comment:
Some other code in this class uses synchronization on `this`, so I think it would be better to synchronize on `sparkContextPromise` in this case.

Reply:
Used `sparkContextPromise` as lock.